runtime/caml/major_gc.h

```diff
@@ -21,21 +21,28 @@
 #include "freelist.h"
 #include "misc.h"
 
+/* An interval of a single object to be scanned.
+   The end pointer must always be one-past-the-end of a heap block,
+   but the start pointer is not necessarily the start of the block */
+typedef struct {
+  value* start;
+  value* end;
+} mark_entry;
+
```

Comment:
`Obj.truncate` carefully leaves the truncated part (even its header) scannable because the ref table(s) may still point into it, so it should be safe. At any rate, truncation is deprecated and incompatible with multicore, so it will have to be removed rather soon.